### ISSUE TITLE
How to catch Errors thrown by request async code?!

### ISSUE THREAD

**Issue by Apollon77** [yagop/node-telegram-bot-api#798]
One of our users got an exception like 
https://github.com/iobroker-community-adapters/ioBroker.telegram/issues/152

When I check your code you simply throw errors in asny methods like https://github.com/yagop/node-telegram-bot-api/blob/master/src/telegram.js#L283 ... it is near to impossible to good catch them :-( What about introducing error events or other way that allows to catch such errors?

Or do I miss something?

**Comment by kamikazechaser** [comment]
The line you quoted above (L283) is specific for the Telegram API. The Telegram API will **always** return a status 200 even if your query is bad, naturally you would expect a 4XX, but not with the Telegram API.

Hence, with the request builder, we have to parse the 200 response from Telegram servers, if its a bad query, we throw the error.

Any other error e.g. lost internet connection, proxy errors, e.t.c. can be caught. The user seems to have lost connection hence the error.

**Comment by Apollon77** [comment]
Exactly. But how to catch them? A try catch around the api call will not do the job as i experienced. So only way i see Is to use the global „uncaugth exception“ handler which is not the best way ...

**Comment by kamikazechaser** [comment]
Maybe we could instead throw individual errors which could be from a known object of Telegram errors?

**Comment by Apollon77** [comment]
But also these would be not catchable unless you use the global exception handler (yes more easy to detect that way and so better then nothing. 
Better would be to allow to provide an on-Error callback somewhere or a promise reject if promises are used or an error event if events are used.

**Comment by yagop** [comment]
Thanks for the detailed write-up back in 2020 - and good news: everything you asked for here is implemented as of the v1.1.0 TypeScript rewrite. There are now typed, catchable errors plus error events, so you no longer have to fall back on the global `uncaughtException` handler.

## Typed errors from a known set
The library throws a small error hierarchy ([`src/errors.ts`](https://github.com/yagop/node-telegram-bot-api/blob/master/src/errors.ts)):

- `FatalError` - code `EFATAL` (network failure / programmer error; carries `.cause`)
- `ParseError` - code `EPARSE` (response wasn't valid JSON; carries `.response`)
- `TelegramError` - code `ETELEGRAM` (the API returned `ok: false`; carries `.response = { status, body }`)

They're exported, and also reachable via `TelegramBot.errors`.

## Catching them on a direct API call (promise rejection)
Every method funnels through `_request()` -> `this.http.request()`, which returns a promise that *rejects* with one of the typed errors above. So a plain `try/catch` (or `.catch()`) just works now:

```js
try {
  await bot.sendMessage(chatId, text);
} catch (err) {
  if (err.code === 'ETELEGRAM') {
    // err.response.body has Telegram's { ok:false, error_code, description }
    console.error('Telegram rejected it:', err.response.body);
  } else if (err.code === 'EFATAL') {
    console.error('Network/transport failure:', err.cause);
  } else {
    throw err;
  }
}

// or, with .catch():
bot.sendMessage(chatId, text).catch(err => { /* err.code, err.response */ });
```

`instanceof` works too, e.g. `err instanceof TelegramBot.errors.TelegramError`.

## Errors from the update sources (events)
Errors thrown by the long-poll loop or the webhook server (where there is no call to `await`) are surfaced as events, so you don't need the global handler:

```js
bot.on('polling_error', (err) => console.error('polling:', err.code, err.message));
bot.on('webhook_error', (err) => console.error('webhook:', err.code, err.message));
```

(If you don't register a listener, the library logs to stderr as a fallback rather than crashing.)

Since the original request - typed/catchable errors plus an error event - is fully covered, I'm going to close this as completed. If you hit a case where an error still escapes to `uncaughtException` in v1.1.0, please open a fresh issue with a minimal repro and we'll dig in. Thanks again!
